crdt.md

```diff
@@ -0,0 +1,25 @@
+The spindexer crdt is designed to support p2p mutations of a replicated dataset.
+
+Mutations are performed on a piece of data using messages. Messages are stored in a central location. At the current state of the spindexer infrastructure, there is only a single spindexer node running and therefore the messages are stored within the spindexer database. Once there are multiple spindexer nodes running, messages will be stored externally using a publicly accessible storage solution, so that all instances of spindexer can read/write to the messages list.
+
+Note: this is not a fully p2p design as it relies on the central storage of the messages list.
+
+
+# Messages
+Messages are the mechanism for mutating data within spinamp. Messages are a G-Set, meaning that messages can only be appended to the messages list.
+
+Messages consist of a timestamp to provide order, the dataset, the field, and the value to mutate. Messages are idempotent, meaning that the same messages can be applied multiple times and result in the same mutations being applied and arriving at the same final state.
+Messages are commutative, in that messages can be applied in any order, but will only be applied if a newer timestamp for the same column and id has not been seen yet. This means messages must be processed using a last write wins (LWW) strategy
```

Comment:
Cool, I think this should continue to work even as we decentralize, because in most cases each (id, table) pair can still have a centralized source of messages. Eg: Artist update messages will come from the artist and always be signed by their single key.

Of course it has problems (eg: an artist could accidentally create a far-in-the-future timestamp and fuck up their record permanently), but we'll improve the design to handle that when it becomes more decentralized.

Reply:
i do think we should flesh out this doc into a more thorough spec though too.